[Chore] added profile.service test class & fixed error that returned undefined (Sentry)

Malyuk-A opened the pull request:
### Checklist
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->

<!-- If it fixes an open issue, please link to the issue here. -->
This PR originated from the following Sentry issue [ARTEMIS-K3](https://sentry.ase.in.tum.de/organizations/artemis/issues/613/?project=2#exception). _(I was assigned to it)_

### Description :
<!-- Describe your changes in detail -->
### New Test File 
I added a new test file `profile.service.spec.ts` for `profile.service.ts`. _(There was no test file for this service before.)_
This test file includes two tests:

1. `'should call correct URL'`
2. `'should get the profile info'`
    - This test case tests the service method `getProfileInfo`, which calls every other method that exists in `profile.service`. _(Thus, testing the entire functionality of the service)_

###  Concrete Service Fix
I refactored the service itself slightly. _(I explain my reasoning for these changes below.)_
Basically, I just made sure to set some properties ( `externalUserManagementName = "` & `externalUserManagementURL`  ) if they are `undefined`.
Additionally, I replaced the return value of the `getProfileInfo` method from `BehaviorSubject` to `Observable`. 
Otherwise, errors occur.
E.g. `getProfileInfo` will return `undefined` in my test because : 
If you call `next()` on a `BehaviorSubject` with `0 Observers` it will return `undefined`, regardless of the value inside of `next(value_x)` and regardless of the starting value of the initialized `BehaviorSubject(init)`. _[(Here is the small experiment)](https://stackblitz.com/edit/rxjs-behaviorsubject-simpleexample-8u934v?file=index.ts)_

![image](https://user-images.githubusercontent.com/65814168/120004129-b37aad00-bfd6-11eb-980c-92326e232979.png)



**Further Explanation :**
We constantly work with `const data = res.body!`, this `res` SEEMS to be of type `HttpResponse<ProfileInfo>`, therefore `data` SHOULD be just of type `ProfileInfo`, which is defined in `profile-info.model.ts`. But it is not.
After doing a `GET` request with [Postman](https://www.postman.com/) on one of the test servers, I finally understood my issue and the goal of `profile.service`.
![image](https://user-images.githubusercontent.com/65814168/120004619-47e50f80-bfd7-11eb-9e83-c169155d8797.png)

It **translates** the (slightly) different profile information received from the server.
E.g. Server : `allowed-minium-orion-version` -> `allowedMinimumOrionVersion` (therefore mapping it to an **actual** `profileInfo` object.).

The service tries to access the following properties of `data`:
- `data['display-ribbon-on-profiles']`
- `data['allowed-minimum-orion-version']`
- `data['test-server']`

These properties are **not** defined in `profile-info.model.ts`. 
But the following are :
- `public allowedMinimumOrionVersion: string`
- `public testServer?: boolean`

This means that the tests do **not** need to insert a true `profileInfo` object but one using the "server-style".
Thus, I reused the received data from Postman for the tests.
 
### Connection to Sentry issue
The issue in a nutshell:
- Problem location : `clone-repo-button.component`
- A TypeError occurs because the property `'sshKeysURL` cannot be read from `undefined`.
- Based on the stack trace and details provided by Sentry the (first/only) location where this issue can occur is in line 59 `this.sshKeysURL = info.sshKeysURL;`

![image](https://user-images.githubusercontent.com/65814168/119222771-4a86c700-baf6-11eb-94b0-541f5114d4a0.png)
As you can see the origin of `info` comes from `profile.service`.
- I figured that the service is creating the problem. Furthermore, this could be the source for a multitude of other (Sentry, and beyond) issues.
- I decided to look into the service to find where the issue might be, why it returns `undefined`.
- I discovered that there was no test file for this service and created one.

**Conclusion :** I hope this small fix of the service will also fix the Sentry Issue (and maybe other problems as well).

### Steps for Testing
Read my description carefully and look at the code.

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->

profile.service.ts          93% lines covered
profile-info.model.ts 100% lines covered

![image](https://user-images.githubusercontent.com/65814168/120005525-32bcb080-bfd8-11eb-823e-76e3909142a7.png)

**Thank you very much for your support** 
- @Olimbia23 
- @SimonEntholzer 
- @iozbeyli 
- @FrankeLukas

krusche commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -73,11 +77,19 @@ export class ProfileService {
+        if (data['allowed-minimum-orion-version']) {
+            profileInfo.allowedMinimumOrionVersion = data['allowed-minimum-orion-version'];
+        } else {
+            profileInfo.allowedMinimumOrionVersion = data['allowedMinimumOrionVersion'];
```
I don't really understand the else clause here? Why should the key change to camel case?

Malyuk-A replied:
I use camel case for this property because there is no such property with the name `allowed-minimum-orion-version` in `profile-info.model.ts`. But there is one in camel case. Therefore the non camel case one will not be found and return undefined. This could lead to an error further on. 

The reason why I want to start and continue supporting the non camel case one (in the 'if-then' statement) is that it seems to be working so far but it does not come from `profile-info.model.ts` natively but was most likely appended somewhere else. If I remove this option other parts of the code might break.
![image](https://user-images.githubusercontent.com/65814168/119275827-882f4100-bc17-11eb-9746-71ea1ad014a5.png)
I hope my reasoning become more clear?
(If I am wrong please correct me.)

krusche replied:
data is coming from the server. It won't have a camelCase style property. That's why we need to convert after all

Malyuk-A replied:
Yes, you are absolutely right, after using Postman to receive an actual server response I understood the issue and the reason for `profile.service`.  Thus, I removed the `if-else` blocks.

krusche commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -73,11 +77,19 @@ export class ProfileService {
+        if (data['test-server']) {
+            profileInfo.testServer = data['test-server'];
+        } else {
+            profileInfo.testServer = data['testServer'];
```
I don't really understand the else clause here? Why should the key change to camel case?

Malyuk-A replied:
Please have a look at my reply above to a similar issue.

Malyuk-A replied:
My fault. (See description or answer above)

krusche commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -26,17 +26,21 @@ export class ProfileService {
-                            const ribbonProfiles = displayRibbonOnProfiles.filter((profile: string) => profileInfo.activeProfiles.includes(profile));
-                            if (ribbonProfiles.length !== 0) {
-                                profileInfo.ribbonEnv = ribbonProfiles[0];
+                            if (data['display-ribbon-on-profiles']) {
```
you use `data['display-ribbon-on-profiles']` multiple times, so it would make sense to use const variable here

Malyuk-A replied:
I checked the code and I use `data['display-ribbon-on-profiles']` exactly 2 times right next to each other. So It does not seem to me to be a big issue when it comes to duplication. 

Or is the issue not about duplication, but about a possible change of the value, which could sabotage the execution? This also seems quite unlikely to me, because the two instances are one after another and a quick check, if the value is empty, should be no big/time-consuming issue.

If you disagree, please tell me so and explain to me why this is indeed an issue.

Olimbia23 commented on src/test/javascript/spec/service/profile.service.spec.ts:
```diff
@@ -0,0 +1,76 @@
+            expect(req.request.url).toEqual(infoUrl);
+        });
+
+        it('should get the profile info', () => {
+            const expected = profileInfo;
+
+            let requestResult = {} as ProfileInfo;
+
+            service.getProfileInfo().subscribe((received) => {
+                if (received) {
+                    requestResult = received;
+                }
+            });
+
+            const req = httpMock.expectOne({ method: 'GET' });
+            req.flush(profileInfo);
+            expect(requestResult).toEqual(expected);
+        });
```
Since the test involves `Observable` subscription I think we should wrap this test into `fakeAsync` block, together with the `tick()` method, as was done in other tests https://confluence.ase.in.tum.de/display/ArTEMiS/Client+Service+Test+Fix

iozbeyli replied:
This also works but using fakeAsync will remove all these if statements which looks a bit odd in my opinion.

Malyuk-A replied:
1.)  Thank you @Olimbia23 for your assumption. I informed myself about the topic and I think you are right. I changed the Behavior Subject to (Publish)Subject.
2.) Due to the changed subject, I no longer have to make sure that `received` is not undefined. (This was my main reason for not using `fakeAsync&Tick`. But now I can use it again, and the code is indeed cleaner and smaller. 👍

Olimbia23 reviewed (commented):
Thank you very much for your detailed explanation! @Malyuk-A 

My assumption is that the problem comes from the `profileInfo` being a `BehaviorSubject` which is initialized with `undefined` at https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/shared/layouts/profiles/profile.service.ts#L21. 
Upon subscription the latest event is replayed, which is `undefined` in our case.
I assume the solution would be to have a regular Subject (`PublishSubject`) instead of Behaviour one.
This would mean that the stream is not of type `ProfileInfo | undefined`, but will always contain `ProfileInfo` data.
However, this is my assumption.

iozbeyli commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -26,17 +26,21 @@ export class ProfileService {
+                            if (data['display-ribbon-on-profiles']) {
+                                const displayRibbonOnProfiles = data['display-ribbon-on-profiles'].split(',');
+                                const ribbonProfiles = displayRibbonOnProfiles.filter((profile: string) => profileInfo.activeProfiles.includes(profile));
+                                if (ribbonProfiles.length !== 0) {
+                                    profileInfo.ribbonEnv = ribbonProfiles[0];
+                                }
```
Shouldn't you set it to '' in an else statement here as well? I don't know the code but it looks like that from the previous version.

Malyuk-A replied:
Yes, I think you are correct. I replaced the else with an undefined check of `ribbonEnv`, if it is indeed undefined I set it to `''`. Thank you for spotting this 👍

Malyuk-A commented:
After changing the Subject (it was BehaviorSubject) multiple Tests fail because they expect a BehaviorSubject with an initial value not always `undefined`, e.g. in `build-plan-link.directive.spec.ts` the `profileInfo = { buildPlanURLTemplate: 'https://some.url.com/plans/{buildPlanId}/path/{projectKey}' }` is inserted. 
But now with just Subject, there is no such option. Or can I insert it in some other way? 

This change of BehaviorSubject to Subject seems to break more than it fixes. Do you think I should continue changing this (BehaviorSubject -> Subject) in other files or simply go back to BehaviorSubject? (@Olimbia23 & @iozbeyli )
Thank you in advance!

Olimbia23 commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -12,31 +12,36 @@ import { Saml2Config } from 'app/home/saml2-login/saml2.config';
     constructor(private http: HttpClient, private featureToggleService: FeatureToggleService) {}
 
-    getProfileInfo(): BehaviorSubject<ProfileInfo | undefined> {
+    getProfileInfo(): Subject<ProfileInfo> {
```
```suggestion
    getProfileInfo(): Observable<ProfileInfo> {
```

Malyuk-A replied:
@Olimbia23 thank you a lot for your continuous support. 

Could you please explain to me why we should use `Observable` here?
Moreover, what does the `OperatorFunction<ProfileInfo | undefined, ProfileInfo>` do exactly?
Can this change, similar to the change to `Subject`, influence other files? (See `about-us.component.spec` where I had to change `BehaviorSubject` to `Observable`. The tests still work. Or is this change safe?)

Thank you in advance!

Olimbia23 replied:
1. Why `Observable`?
Because the `pipe` method returns an `Observable<T>`

2. What does `OperatorFunction<ProfileInfo | undefined, ProfileInfo>`?
Without this cast operator, the return type would be either `Observable<ProfileInfo | undefined>`, but since our method `getProfileInfo()` should return only data types of `Observable<ProfileInfo>`, then we need to use the casting operator. Our Filter method makes sure that we are not returning any undefined data.

3. Can this influence other files?
Yes, because the API is changed (returned type of `getProfileInfo` function). 
Before it was returning `BehaviorSubject` where one can push new elements into by calling the `next` function. `Observable`'s do not have the next function and therefore can not be mutated.
However, it seems like all usages of this function (outside of tests) are just subscribing to the the `Observable` stream without writing anything into it via the `next` method, which is only available for `BehaviorSubject`. Therefore this is not a breaking change (outside of tests, which have to be adjusted accordingly)

Malyuk-A replied:
Thank you once more for your great help and detailed explanations!

Olimbia23 commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -3,7 +3,7 @@ import { HttpClient, HttpResponse } from '@angular/common/http';
 import { SERVER_API_URL } from 'app/app.constants';
 import { ProfileInfo } from './profile-info.model';
-import { BehaviorSubject } from 'rxjs';
+import { Subject } from 'rxjs';
```
```suggestion
import { BehaviorSubject, Observable, OperatorFunction } from 'rxjs';
import { map, filter } from 'rxjs/operators'
```

Malyuk-A replied:
I changed the imports.

Olimbia23 reviewed (commented):
@Malyuk-A We can go back to BehaviorSubject but this time try to filter the events that are `undefined`.
Therefore you can try replacing https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/shared/layouts/profiles/profile.service.ts#L72 with:
`return this.profileInfo.pipe(filter(x => x !== undefined) as OperatorFunction<ProfileInfo | undefined, ProfileInfo>);`

Olimbia23 commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -12,31 +12,36 @@ import { Saml2Config } from 'app/home/saml2-login/saml2.config';
 export class ProfileService {
     private infoUrl = SERVER_API_URL + 'management/info';
-    private profileInfo: BehaviorSubject<ProfileInfo | undefined>;
+    private profileInfo: Subject<ProfileInfo>;
```
this can be reverted as it was before.

Malyuk-A replied:
Has been reverted.

Olimbia23 commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -12,31 +12,36 @@ import { Saml2Config } from 'app/home/saml2-login/saml2.config';
+    getProfileInfo(): Subject<ProfileInfo> {
         if (!this.profileInfo) {
-            this.profileInfo = new BehaviorSubject(undefined);
+            this.profileInfo = new Subject();
```
this also can be reverted.

Malyuk-A replied:
Is reverted.

Malyuk-A commented:
Sorry for the other file changes, I am currently working on removing this mix-up. (Too many branches at the same time, etc...)

Malyuk-A commented:
I refactored my changed and PR description.

The only thing currently that is not working is `course-update.component.spec.ts:128:83 - error TS2345: Argument of type 'BehaviorSubject<ProfileInfo | undefined>' is not assignable to parameter of type 'Observable<ProfileInfo>'.`

I changed the file to fix this error, but there is another error that blocks me from committing or pushing my change.

I only changed a few imports and one line of code. 
![image](https://user-images.githubusercontent.com/65814168/120007646-6ac4f300-bfda-11eb-9b04-e7a13828ba73.png)

But **even if I revert the changes**, there is an error, that I struggle to understand and therefore fix.
It appears in the very last line and stays regardless of the file content (I tried removing different parts of the code, it only disappears if the file is completely empty).
![image](https://user-images.githubusercontent.com/65814168/120007873-ad86cb00-bfda-11eb-8b2c-e28fe2d1de6e.png)

Olimbia23 commented on src/test/javascript/spec/component/about-us/about-us.component.spec.ts:
```diff
@@ -44,7 +44,7 @@ describe('AboutUsComponent', () => {
         const getStaticJsonFromArtemisServerStub = sandbox.stub(staticContentService, 'getStaticJsonFromArtemisServer').returns(of(new AboutUsModel([], [])));
         const getProfileInfoSub = sandbox.stub(profileService, 'getProfileInfo');
-        getProfileInfoSub.returns(new BehaviorSubject<ProfileInfo | undefined>({ inProduction: false, sshCloneURLTemplate: 'ssh://git@testserver.com:1234/' } as ProfileInfo));
+        getProfileInfoSub.returns(new Observable<ProfileInfo>());
```
```suggestion
        getProfileInfoSub.returns(new BehaviorSubject<ProfileInfo>({ inProduction: false, sshCloneURLTemplate: 'ssh://git@testserver.com:1234/' } as ProfileInfo).asObservable());
```

This seems to change the existing logic. Before it was returning a `BehaviorSubject` with a predefined object and now it returns an empty `Observable` stream.

Can you try to invoke `.asObservable()` on this `BehaviorSubject` and should return `Observable<T>`
https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/asobservable.md#rxobservableasobservable-%E2%93%A2

I would assume that this should help and make the tests pass.
I guess this change has to be applied on every breaking test.

Malyuk-A replied:
Yes, indeed this seems to do the trick.
(This is more intuitive and handy than trying to insert a starting value in an Observable with `startWith()`)
Great, thanks for sharing this knowledge👍

Moreover, the error that I had with `course-update.component.spec.ts` has disappeared. 
So now everything should be working.

MDK5 reviewed (changes requested):
Found a small suggestion for an improvement.

Please update the PR description. Several mentioned changes are not present in the code (anymore perhaps), e.g. renaming variables.

Also I'm not sure if the conversion to Observable really solves the issue. While filtering undefined surely prevents undefined from being emitted if some component subscribes before the initialization call, the conversion makes it lose the imporant property of emitting the current value on a new subscription. If I understand it correctly, any component subscribing after the initialization took place will now never receive the data at all, which will very likely cause other problems. Information from [this SO question](https://stackoverflow.com/questions/39494058/behaviorsubject-vs-observable), especially [this answer](https://stackoverflow.com/a/53355864/11948496). A ReplaySubject seems to be the variation with all needed properties.

Additionally it would be great if you could explain the steps needed to reproduce the problem to it can be tested if it is fixed.

Olimbia23 commented:
> Found a small suggestion for an improvement.
> 
> Please update the PR description. Several mentioned changes are not present in the code (anymore perhaps), e.g. renaming variables.
> 
> Also I'm not sure if the conversion to Observable really solves the issue. While filtering undefined surely prevents undefined from being emitted if some component subscribes before the initialization call, the conversion makes it lose the imporant property of emitting the current value on a new subscription. If I understand it correctly, any component subscribing after the initialization took place will now never receive the data at all, which will very likely cause other problems. Information from [this SO question](https://stackoverflow.com/questions/39494058/behaviorsubject-vs-observable), especially [this answer](https://stackoverflow.com/a/53355864/11948496). A ReplaySubject seems to be the variation with all needed properties.
> 
> Additionally it would be great if you could explain the steps needed to reproduce the problem to it can be tested if it is fixed.

Thank you for your references to SO @MDK5. From my understanding, we are not not using raw `Observables` in our case. We do not create them via `Observable.of` or `create` or `from` methods, but rather we are casting it from the `BehaviorSubject`, which preserves its properties.
I tried a simple example via the sandbox and it seems like current value is still replayed on a new subscription.
Could you please confirm? Link to the example sandbox: https://jsfiddle.net/75go1de0/2/

Olimbia23 reviewed (approved):
The changes look good to me. 
Thank you, @Malyuk-A for your efforts on this PR!

Malyuk-A commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -41,14 +41,22 @@ export class ProfileService {
+
+                        // Both properties below are mandatory Strings (see profile-info.model.ts)
+                        const externalUserManagementName = data.externalUserManagementName;
+                        profileInfo.externalUserManagementName = externalUserManagementName === undefined ? '' : externalUserManagementName;
```
👋 Hey Martin ,

### A) The Nullish coalescing operator ??
Thank you for showing me this operator. It is new to me and seems to be quite handy.
If I am not mistaken I can shorten my code even more:
Instead of e.g. 
`const externalUserManagementName = data.externalUserManagementName;`
`profileInfo.externalUserManagementName = externalUserManagementName === undefined ? '' : externalUserManagementName;`
Simply this one line:
`profileInfo.externalUserManagementName = data.externalUserManagementName ?? '';`
I hope this is correct?

### B) About recreating the problem
![image](https://user-images.githubusercontent.com/65814168/120099421-9dd2c800-c13b-11eb-9ca8-745a31fc72d4.png)
![image](https://user-images.githubusercontent.com/65814168/120099427-a4f9d600-c13b-11eb-868d-b483ebc6bea0.png)
I already talked with @SimonEntholzer about this. It seems to be a very niche problem (just think about the number of times students copy a repository or use the profile.service, this is basically used for every programming exercise and log in, so the numbers could be way bigger. Additionally, I do not touch the file where Sentry catches this issue.
 `(clone-repo-button.component)`
Moreover, I think this PR is big and old enough. If this issue continues to reappear I will look into it once more (in a new PR). 

----
I will look over my description again and adjust it 👍

MDK5 reviewed (dismissed):
@Olimbia23 I stand corrected, your example certainly does work. I'm sorry for wasting your time, I should've tried it out before raising the question.

Code changes look good to me :+1:

SimonEntholzer reviewed (approved):
The  test makes sense :+1:. The changes to the service seem reasonable too.

FrankeLukas commented on src/test/javascript/spec/service/profile.service.spec.ts:
```diff
@@ -0,0 +1,246 @@
+                    tutor_assessment_tour: 'Patterns in Software Engineering',
+                },
+            },
+            //            "course-group-students": "artemis-artemistutorial-students"
```
Should this code be removed? (Dead code)

Malyuk-A replied:
Yes, my mistake.

FrankeLukas commented on src/test/javascript/spec/service/profile.service.spec.ts:
```diff
@@ -0,0 +1,246 @@
+        testServer: true,
+        ribbonEnv: '',
+        guidedTourMapping: {
+            //         "course-group-students": "artemis-artemistutorial-students",
```
Same here.

FrankeLukas commented on src/main/webapp/app/shared/layouts/profiles/profile.service.ts:
```diff
@@ -69,7 +74,7 @@ export class ProfileService {
         }
 
-        return this.profileInfo;
+        return this.profileInfo.pipe(filter((x) => x !== undefined) as OperatorFunction<ProfileInfo | undefined, ProfileInfo>);
```
We should use `!=` here to filter `null` as well: https://www.tektutorialshub.com/typescript/typescript-null-undefined-strict-null-checks/

Malyuk-A replied:
Yes, you are correct. Thank you for the provided article.

FrankeLukas reviewed (dismissed):
Lgtm otherwise.
